animation-proposal.md

```diff
@@ -0,0 +1,338 @@
+
+In short:
+- Add support for playing animations defined on assets (loaded from glTF files, Unity asset bundles, etc.).
+- Add generalized support for interpolating any mutable property of an `Actor` or `Asset`.
+- Remove support for programmatic animaton generation.
+- Remove the existing `animateTo` call in favor of the new interpolation solution.
+
+
+
+## API
+
+### Animations
+
+#### Move animation API to new class called `Animation`
+
+When a model is loaded from a resource, its animations will be included in the `object-spawned` reply, and be available on `Actor` as an array of `AnimationLike` types. `AnimationLike` will define the API for starting, stopping, seeking, etc. animations.
+
+```ts
+export interface AnimationLike {
+	/**
+	 * getter that returns the name of the animation.
+	 */
+	public name: string;
+
+	/**
+	 * Starts playing the animation from the beginning.
+	 * @returns A Promise that resolves when the animation completes or is stopped.
+	 */
+	public start(): Promise<void>;
+
+	/**
+	 * Stops the animation on the actor and resolves promises returned from `completed()`.
+	 */
+	public stop();
+
+	/**
+	 * Pauses the animation (sets animation speed to zero).
+	 */
+	public pause();
+
+	/**
+	 * Resumes the animation (sets animation speed to 1).
+	 */
+	public resume();
+
+	/**
+	 * Sets the animation time (units are in seconds).
+	 */
+	public setTime(time: number);
+
+	/**
+	 * Sets the playback speed.
+	 */
+	public setSpeed(speed: number);
+
+	/**
+	 * (Advanced) Sets the time, speed, and enabled state of an animation.
+	 */
+	public setState(state: SetAnimationStateOptions);
+
+	/**
+	 * @returns A Promise that resolves when the animation completes or is stopped.
+	 */
+	public completed(): Promise<void>;
+}
+```
+
+Then, on `Actor`:
+
+```ts
+/**
+ * getter that returns the array of animations defined on this actor.
+ */
+public get animations(): AnimationLike[];
+
+/**
+ * Get an animation by name.
+ */
+public animation(animationName: string): AnimationLike;
+```
+
+#### Examples
+
+```ts
+// Play the 'jump' animation, then do something when it's done.
+actor.animation("jump").start().then(() => console.log("done jumping"));
+
+// Play the 'victory' animation, and wait for it to complete.
+await actor.animation("victory").start();
+```
+
+### Interpolations
+
+#### Add `Context.interpolate` method
+
+Add an `interpolate` method to class `Context`. Calling `interpolate` initiates a lerp operation on the host and returns a Promise that will resolve after `seconds` time has elapsed or the interpolation is otherwise canceled.
+
+```ts
+/**
+ * Interpolates the field indicated by `selector` to `value` over `interval` seconds along `options.curve`.
+ * @param selector The selector of the field to interpolate. Get this value by calling actor.selector(fieldpath).
+ * @param value The value to interpolate to or by, depending on the value of options.basis. Can be a selector or a concrete value.
```

Comment:
Recommend naming this `to`, and adding an optional `from` argument, in case you want to reset the value and then animate it.

Reply:
I like the renaming, but unsure how `to` could be optional, since interpolations need a target value. Can you expand?

Reply:
I see now. Good suggestion.